packages/react-components/src/components/grid.tsx

```diff
@@ -51,162 +46,175 @@ type State = {
+    const unmounted = useRef(false)
 
-    componentDidMount() {
-        this.unmounted = false
-        this.onFetch()
-    }
+    const handleFetchGifs = useCallback(
+        (prefetchCount: number) => {
+            const debounceFetchGifs = debounce(FETCH_DEBOUNCE, async (prefetchCount: number) => {
+                let gifs
+                try {
+                    gifs = await paginator.current()
+                    if (unmounted.current) return
+                } catch (e) {
+                    const error = e as Error
+                    if (unmounted.current) return
+                    dispatch({ type: 'FETCH_FAILURE', error })
+                    if (onGifsFetchError) onGifsFetchError(error)
+                    return
+                }
 
-    componentWillUnmount() {
-        this.unmounted = true
-    }
+                if (gifs) {
+                    if (prefetchCount === gifs.length) {
+                        dispatch({ type: 'DONE_FETCHING' })
+                    } else {
+                        dispatch({ type: 'FETCH_SUCCESS', gifs })
+                        if (onGifsFetched) onGifsFetched(gifs)
+                    }
+                }
+            })
 
-    onLoaderVisible = (isVisible: boolean) => {
-        if (this.unmounted) return
-        this.setState({ isLoaderVisible: isVisible }, this.onFetch)
-    }
+            if (unmounted.current) return
 
-    fetchGifs = debounce(Grid.fetchDebounce, async (prefetchCount) => {
-        let gifs
-        try {
-            gifs = await this.paginator()
-            if (this.unmounted) return
-        } catch (error) {
-            if (this.unmounted) return
-            this.setState({ isFetching: false, isError: true })
-            const { onGifsFetchError } = this.props
-            if (onGifsFetchError) onGifsFetchError(error as Error)
-        }
-        if (gifs) {
-            // if we've just fetched and we don't have
-            // any more gifs, we're done fetching
-            if (prefetchCount === gifs.length) {
-                this.setState({ isDoneFetching: true })
-            } else {
-                this.setState({ gifs, isFetching: false })
-                const { onGifsFetched } = this.props
-                if (onGifsFetched) onGifsFetched(gifs)
-                this.onFetch()
+            debounceFetchGifs(prefetchCount)
+        },
+        [onGifsFetched, onGifsFetchError]
+    )
+
+    const onLoaderVisible = useCallback(
+        (isVisible: boolean) => {
+            if (unmounted.current) return
+            dispatch({ type: 'LOADER_VISIBLE', isVisible })
+            if (!state.isFetching && isVisible) {
+                dispatch({ type: 'START_FETCH' })
+                handleFetchGifs(state.gifs.length)
             }
+        },
+        [state.gifs.length, state.isFetching, handleFetchGifs]
+    )
+
+    useEffect(() => {
+        unmounted.current = false
+        if (unmounted.current === false) {
```

Comment:
@matheusmorett2 isn't this always true since you're setting the ref right above?